trino-s3-proxy/src/main/java/io/trino/s3/proxy/server/minio/Signer.java

```diff
@@ -0,0 +1,444 @@
+import java.util.Set;
+import java.util.TreeMap;
+import java.util.stream.Collectors;
+
+/**
+ * Amazon AWS S3 signature V4 signer.
+ */
+public class Signer
+{
+    //
+    // Excerpts from @lsegal - https://github.com/aws/aws-sdk-js/issues/659#issuecomment-120477258
+    //
+    // * User-Agent
+    // This is ignored from signing because signing this causes problems with generating pre-signed
+    // URLs (that are executed by other agents) or when customers pass requests through proxies, which
+    // may modify the user-agent.
+    //
+    // * Authorization
+    // Is skipped for obvious reasons.
+    //
+    // * Accept-Encoding
+    // Some S3 servers like Hitachi Content Platform do not honour this header for signature
+    // calculation.
+    //
+    private static final Set<String> IGNORED_HEADERS =
+            ImmutableSet.of("accept-encoding", "authorization", "user-agent");
+    private static final Set<String> PRESIGN_IGNORED_HEADERS =
+            ImmutableSet.of(
+                    "accept-encoding",
+                    "authorization",
+                    "user-agent",
+                    "content-md5",
+                    "x-amz-content-sha256",
+                    "x-amz-date",
+                    "x-amz-security-token");
+
+    private MinioRequest request;
+    private String contentSha256;
+    private ZonedDateTime date;
+    private String region;
+    private String accessKey;
+    private String secretKey;
+    private String prevSignature;
+
+    private String scope;
+    private Map<String, String> canonicalHeaders;
+    private String signedHeaders;
+    private MinioUrl url;
+    private String canonicalQueryString;
+    private String canonicalRequest;
+    private String canonicalRequestHash;
+    private String stringToSign;
+    private byte[] signingKey;
+    private String signature;
+    private String authorization;
+
+    /**
+     * Create new Signer object for V4.
+     *
+     * @param request HTTP Request object.
+     * @param contentSha256 SHA-256 hash of request payload.
+     * @param date Date to be used to sign the request.
+     * @param region Amazon AWS region for the request.
+     * @param accessKey Access Key string.
+     * @param secretKey Secret Key string.
+     * @param prevSignature Previous signature of chunk upload.
+     */
+    private Signer(
+            MinioRequest request,
+            String contentSha256,
+            ZonedDateTime date,
+            String region,
+            String accessKey,
+            String secretKey,
+            String prevSignature)
+    {
+        this.request = request;
+        this.contentSha256 = contentSha256;
+        this.date = date;
+        this.region = region;
+        this.accessKey = accessKey;
+        this.secretKey = secretKey;
+        this.prevSignature = prevSignature;
+    }
+
+    private void setScope(String serviceName)
+    {
+        this.scope =
+                this.date.format(Time.SIGNER_DATE_FORMAT)
```

Comment:
please fix all of these short line breaks.  minio likely uses a crazy narrow style